### PR TITLE
feat: add autotune to optimize attention speed

### LINKED ISSUE #114: Evaluate different block size in attention kernel
In the current triton attention kernel implementation, we use the same value for BLOCK_N and BLOCK_M

```
     BLOCK_M=BLOCK_M,
     BLOCK_N=BLOCK_M,
```
- We should evaluate the faisability of this change (if non square matrix are supported)
- We should evaluate the performance difference of computing a different block size for BLOCK_N.

### PR DESCRIPTION
**Pull request by pommedeterresautee**
fix #114

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
test pass

```log

============================================================================================== 2729 passed in 8320.20s (2:18:40) ==============================================================================================

```

**Comment by pommedeterresautee** [comment]
dynamo tests pass after refactoring (autotune and new shapes)

```
pytest test/test_torchdynamo.py -x 
=================================================================================================== 266 passed in 4729.95s (1:18:49) ===================================================================================================
```

File: src/kernl/implementations/attention.py [hunk 1]
```diff
@@ -274,21 +348,23 @@ def _fwd_kernel(
...
-            offs_mask = offs_base_mask + offs_n[None, :] * attention_mask_k_stride
-            if MASK_M_SIZE != 1:  # mask has a square shape, we load (BLOCK_M, BLOCK_N) elements
+            offs_mask = offs_base_mask + offs_n[None, :] * attention_mask_n_stride
+            if IS_MATRIX_MASK:  # mask has a matrix shape, we load (BLOCK_M, BLOCK_N) elements
                 offs_mask += offs_m[:, None] * attention_mask_m_stride
 
-            if NEED_LOAD_MASK_SIZE_N & MASK_M_SIZE == 1:  # mask has a vector shape need a load mask
-                attention_load_mask = offs_n[None, :] < size_n
-            if MASK_M_SIZE != 1:  # mask has a matrix shape
+            if NEED_LOAD_MASK_SIZE_N & (not IS_MATRIX_MASK):  # mask has a vector shape + need a load mask
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:355]: `(not IS_MATRIX_MASK) & NEED_LOAD_MASK_SIZE_N`

File: src/kernl/implementations/attention.py [hunk 2]
```diff
@@ -274,21 +348,23 @@ def _fwd_kernel(
...
-                    attention_load_mask = offs_n[None, :] < size_n
+                    attention_load_mask = offs_n[None, :] < attention_mask_n_size
                 elif NEED_LOAD_MASK_SIZE_M & NEED_LOAD_MASK_SIZE_N:  # load mask on both axis
-                    attention_load_mask = (offs_n[None, :] < size_n) & (offs_m[:, None] < size_m)
+                    attention_load_mask = (offs_n[None, :] < attention_mask_n_size) & (
+                        offs_m[:, None] < attention_mask_m_size
+                    )
 
-            if NEED_LOAD_MASK_SIZE_M | NEED_LOAD_MASK_SIZE_N:
+            if (NEED_LOAD_MASK_SIZE_M & IS_MATRIX_MASK) | NEED_LOAD_MASK_SIZE_N:
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:367]: manually checked, minimal condition indeed

File: src/kernl/implementations/attention.py [hunk 3]
```diff
@@ -437,53 +496,33 @@ def forward(
...
+            size_n // 32,  # size_n_cache_key
+            q,  # Q
+            k,  # K
+            v,  # V
+            sm_scale,  # sm_scale
+            attention_mask,  # attention_mask
+            tmp,  # TMP
+            output,  # output
+            *q.stride(),  # (batch, heads, size_m, size_k)
+            *k.stride(),  # # (batch, heads, size_n, size_k)
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:515]: double #

File: src/kernl/implementations/attention.py [hunk 4]
```diff
@@ -437,53 +496,33 @@ def forward(
...
+        _fwd_kernel[grid](  # can't use name args because of the way autotune is implemented :-(
+            heads,  # heads
+            size_m,  # size_m
+            size_n,  # size_n
+            size_m // 32,  # size_m_cache_key
+            size_n // 32,  # size_n_cache_key
+            q,  # Q
+            k,  # K
+            v,  # V
+            sm_scale,  # sm_scale
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:510]: why do you put useless comments when the name match parameter

File: src/kernl/implementations/attention.py [hunk 5]
```diff
@@ -61,12 +61,69 @@ def attention_reference(
     return ref_out
 
 
+def closest_power_of_2(n) -> (int, int):
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:64]: not reviewed, will change

File: src/kernl/implementations/attention.py (hunk 4 above)
Comment (pommedeterresautee) [reply src/kernl/implementations/attention.py:510]: Homogeneous

File: src/kernl/implementations/attention.py (hunk 3 above)
Comment (pommedeterresautee) [reply src/kernl/implementations/attention.py]: done

File: src/kernl/implementations/attention.py (hunk 5 above)
Comment (pommedeterresautee) [reply src/kernl/implementations/attention.py]: modified

**Review by gaetansnl — APPROVED** [review]

**Pull request merged**